File: packages/core/tests/auto-detail.test.ts
/**
 * Tests de `buildAutoDetailComposition` — KRO-73 (B+).
 *
 * Sección canónica → composition esperada. Cubre todas las heurísticas
 * + edge cases (sección vacía, fields que no encajan en ningún slot).
 */

import { describe, it, expect } from 'vitest';
import { buildAutoDetailComposition, buildAutoListComposition } from '../src/auto-detail';
import type { FieldDefLike } from '../src/types';

describe('buildAutoDetailComposition — defaults base', () => {
  it('sección vacía → composition con hero_protagonico + slots vacíos', () => {
    const result = buildAutoDetailComposition([]);
    expect(result.recipe).toBe('hero_protagonico');
    expect(result.action).toBe('none');
    expect(result.slots).toEqual({});
  });
});

describe('buildAutoDetailComposition — avatar / banner', () => {
  it('field con behavior=avatar → slot avatar', () => {
    const fields: FieldDefLike[] = [
      { key: 'pic', type: 'image', behavior: 'avatar' },
    ];
    const r = buildAutoDetailComposition(fields);
    expect(r.slots.avatar).toEqual({ fields: ['pic'] });
  });

  it('field con behavior=banner → slot banner', () => {
    const r = buildAutoDetailComposition([
      { key: 'header', type: 'image', behavior: 'banner' },
    ]);
    expect(r.slots.banner).toEqual({ fields: ['header'] });
  });

  it('field con behavior=cover (sin banner) → slot banner (fallback)', () => {
    const r = buildAutoDetailComposition([
      { key: 'cover', type: 'image', behavior: 'cover' },
    ]);
    expect(r.slots.banner).toEqual({ fields: ['cover'] });
  });

  it('field con behavior=thumbnail → slot banner (fallback)', () => {
    const r = buildAutoDetailComposition([
      { key: 'thumb', type: 'image', behavior: 'thumbnail' },
    ]);
    expect(r.slots.banner).toEqual({ fields: ['thumb'] });
  });

  it('banner explícito gana sobre cover', () => {
    const r = buildAutoDetailComposition([
      { key: 'cover',  type: 'image', behavior: 'cover' },
      { key: 'header', type: 'image', behavior: 'banner' },
    ]);
    expect(r.slots.banner).toEqual({ fields: ['header'] });
  });
});

describe('buildAutoDetailComposition — title', () => {
  it('primer text-short → slot title', () => {
    const r = buildAutoDetailComposition([
      { key: 'nombre', type: 'text' },
    ]);
    expect(r.slots.title).toEqual({ fields: ['nombre'] });
  });

  it('select también es text-short válido', () => {
    const r = buildAutoDetailComposition([
      { key: 'rareza', type: 'select', options: ['raro', 'común'] },
    ]);
    expect(r.slots.title).toEqual({ fields: ['rareza'] });
  });

  it('text con behavior=url NO va a title', () => {
    const r = buildAutoDetailComposition([
      { key: 'web', type: 'text', behavior: 'url' },
    ]);
    expect(r.slots.title).toBeUndefined();
  });

  it('text con behavior=email NO va a title', () => {
    const r = buildAutoDetailComposition([
      { key: 'mail', type: 'text', behavior: 'email' },
    ]);
    expect(r.slots.title).toBeUndefined();
  });
});

describe('buildAutoDetailComposition — subtitle', () => {
  it('primer year → subtitle', () => {
    const r = buildAutoDetailComposition([
      { key: 'temporada', type: 'number', behavior: 'year' },
    ]);
    expect(r.slots.subtitle).toEqual({ fields: ['temporada'] });
  });

  it('primer iso_date → subtitle', () => {
    const r = buildAutoDetailComposition([
      { key: 'fecha', type: 'text', behavior: 'iso_date' },
    ]);
    expect(r.slots.subtitle).toEqual({ fields: ['fecha'] });
  });
});

describe('buildAutoDetailComposition — stats', () => {
  it('un number → stats con 1 field', () => {
    const r = buildAutoDetailComposition([
      { key: 'edad', type: 'number' },
    ]);
    expect(r.slots.stats).toEqual({
      fields:      ['edad'],
      orientation: 'horizontal',
      separator:   ' | ',
    });
  });

  it('5 numbers → stats con primeros 4 (max)', () => {
    const r = buildAutoDetailComposition([
      { key: 'a', type: 'number' },
      { key: 'b', type: 'number' },
      { key: 'c', type: 'number' },
      { key: 'd', type: 'number' },
      { key: 'e', type: 'number' },
    ]);
    expect(r.slots.stats?.fields).toEqual(['a', 'b', 'c', 'd']);
  });

  it('cero numbers → sin slot stats', () => {
    const r = buildAutoDetailComposition([
      { key: 'nombre', type: 'text' },
    ]);
    expect(r.slots.stats).toBeUndefined();
  });
});

describe('buildAutoDetailComposition — body', () => {
  it('textarea → body', () => {
    const r = buildAutoDetailComposition([
      { key: 'desc', type: 'textarea' },
    ]);
    expect(r.slots.body).toEqual({ fields: ['desc'] });
  });

  it('behavior=markdown → body', () => {
    const r = buildAutoDetailComposition([
      { key: 'doc', type: 'text', behavior: 'markdown' },
    ]);
    expect(r.slots.body).toEqual({ fields: ['doc'] });
  });
});

describe('buildAutoDetailComposition — gallery', () => {
  it('array<image> → gallery', () => {
    const r = buildAutoDetailComposition([
      { key: 'fotos', type: 'array<image>' },
    ]);
    expect(r.slots.gallery).toEqual({ fields: ['fotos'] });
  });

  it('behavior=gallery → gallery', () => {
    const r = buildAutoDetailComposition([
      { key: 'g', type: 'array<image>', behavior: 'gallery' },
    ]);
    expect(r.slots.gallery).toEqual({ fields: ['g'] });
  });
});

describe('buildAutoDetailComposition — related', () => {
  it('behavior=card_index_list → related', () => {
    const r = buildAutoDetailComposition([
      { key: 'refs', type: 'array<number>', behavior: 'card_index_list' },
    ]);
    expect(r.slots.related).toEqual({ fields: ['refs'] });
  });
});

describe('buildAutoDetailComposition — composition completa', () => {
  it('sección rica → todos los slots se llenan correctamente', () => {
    const r = buildAutoDetailComposition([
      { key: 'avatar',  type: 'image',         behavior: 'avatar' },
      { key: 'nombre',  type: 'text' },
      { key: 'year',    type: 'number',        behavior: 'year' },
      { key: 'edad',    type: 'number' },
      { key: 'rating',  type: 'number' },
      { key: 'desc',    type: 'textarea' },
      { key: 'fotos',   type: 'array<image>',  behavior: 'gallery' },
    ]);
    expect(r.recipe).toBe('hero_protagonico');
    expect(r.slots.avatar?.fields).toEqual(['avatar']);
    expect(r.slots.title?.fields).toEqual(['nombre']);
    expect(r.slots.subtitle?.fields).toEqual(['year']);
    // year es type=number con behavior=year — pero buildAutoDetail filtra
    // stats por type=number sin importar behavior. ¿Year cuenta?
    // Re-leyendo: stats hace filter(f => f.type === 'number' && f.key !== titleKey).
    // year cuenta (es number, no es title). edad, rating también.
    expect(r.slots.stats?.fields).toEqual(['year', 'edad', 'rating']);
    expect(r.slots.body?.fields).toEqual(['desc']);
    expect(r.slots.gallery?.fields).toEqual(['fotos']);
  });
});

describe('buildAutoListComposition — recipe de LISTA (kind:list, nunca hero)', () => {
  it('sección vacía → row_text + slots vacíos + action none', () => {
    const r = buildAutoListComposition([]);
    expect(r.recipe).toBe('row_text');
    expect(r.action).toBe('none');
    expect(r.slots).toEqual({});
  });

  it('con imagen (avatar) → compact_avatar (NO hero_protagonico)', () => {
    const r = buildAutoListComposition([
      { key: 'pic',    type: 'image', behavior: 'avatar' },
      { key: 'nombre', type: 'text' },
    ]);
    expect(r.recipe).toBe('compact_avatar');
    expect(r.slots.avatar).toEqual({ fields: ['pic'] });
    expect(r.slots.title).toEqual({ fields: ['nombre'] });
  });

  it('image plano (sin behavior) también dispara compact_avatar', () => {
    const r = buildAutoListComposition([
      { key: 'foto',   type: 'image' },
      { key: 'nombre', type: 'text' },
    ]);
    expect(r.recipe).toBe('compact_avatar');
    expect(r.slots.avatar).toEqual({ fields: ['foto'] });
  });

  it('sin imagen → row_text con title + subtitle', () => {
    const r = buildAutoListComposition([
      { key: 'nombre', type: 'text' },
      { key: 'anio',   type: 'number', behavior: 'year' },
    ]);
    expect(r.recipe).toBe('row_text');
    expect(r.slots.title).toEqual({ fields: ['nombre'] });
    expect(r.slots.subtitle).toEqual({ fields: ['anio'] });
  });

  it('NUNCA devuelve una recipe kind:detail (hero/editorial/momento)', () => {
    const r = buildAutoListComposition([
      { key: 'banner', type: 'image', behavior: 'banner' },
      { key: 'titulo', type: 'text' },
      { key: 'cuerpo', type: 'textarea' },
    ]);
    expect(['compact_avatar', 'compact_card', 'row_text']).toContain(r.recipe);
    expect(r.recipe).not.toBe('hero_protagonico');
  });

  it('title y subtitle no colisionan (subtitle != title)', () => {
    const r = buildAutoListComposition([
      { key: 'nombre', type: 'text' },  // title
    ]);
    expect(r.slots.title).toEqual({ fields: ['nombre'] });
    expect(r.slots.subtitle).toBeUndefined();
  });
});

File: packages/core_dart/test/auto_detail_test.dart
/// Corpus de `buildAutoDetailComposition` — ESPEJO 1:1 de `tests/auto-detail.test.ts`.
import 'package:test/test.dart';
import 'package:kromia_core/kromia_core.dart';

FieldDefLike fd(String key, String type, [String? behavior, List<String>? options]) =>
    FieldDefLike(key: key, type: type, behavior: behavior, options: options);

void main() {
  group('buildAutoDetailComposition — defaults base', () {
    test('sección vacía → hero_protagonico + slots vacíos', () {
      final r = buildAutoDetailComposition([]);
      expect(r.recipe, 'hero_protagonico');
      expect(r.action, 'none');
      expect(r.slots, isEmpty);
    });
  });

  group('buildAutoDetailComposition — avatar / banner', () {
    test('behavior=avatar → slot avatar', () {
      final r = buildAutoDetailComposition([fd('pic', 'image', 'avatar')]);
      expect(r.slots['avatar']!.fields, ['pic']);
    });
    test('behavior=banner → slot banner', () {
      final r = buildAutoDetailComposition([fd('header', 'image', 'banner')]);
      expect(r.slots['banner']!.fields, ['header']);
    });
    test('behavior=cover → banner (fallback)', () {
      final r = buildAutoDetailComposition([fd('cover', 'image', 'cover')]);
      expect(r.slots['banner']!.fields, ['cover']);
    });
    test('behavior=thumbnail → banner (fallback)', () {
      final r = buildAutoDetailComposition([fd('thumb', 'image', 'thumbnail')]);
      expect(r.slots['banner']!.fields, ['thumb']);
    });
    test('banner explícito gana sobre cover', () {
      final r = buildAutoDetailComposition([
        fd('cover', 'image', 'cover'),
        fd('header', 'image', 'banner'),
      ]);
      expect(r.slots['banner']!.fields, ['header']);
    });
  });

  group('buildAutoDetailComposition — title', () {
    test('primer text-short → title', () {
      final r = buildAutoDetailComposition([fd('nombre', 'text')]);
      expect(r.slots['title']!.fields, ['nombre']);
    });
    test('select también vale', () {
      final r = buildAutoDetailComposition([fd('rareza', 'select', null, ['raro', 'común'])]);
      expect(r.slots['title']!.fields, ['rareza']);
    });
    test('text con url NO va a title', () {
      final r = buildAutoDetailComposition([fd('web', 'text', 'url')]);
      expect(r.slots['title'], isNull);
    });
    test('text con email NO va a title', () {
      final r = buildAutoDetailComposition([fd('mail', 'text', 'email')]);
      expect(r.slots['title'], isNull);
    });
  });

  group('buildAutoDetailComposition — subtitle', () {
    test('primer year → subtitle', () {
      final r = buildAutoDetailComposition([fd('temporada', 'number', 'year')]);
      expect(r.slots['subtitle']!.fields, ['temporada']);
    });
    test('primer iso_date → subtitle', () {
      final r = buildAutoDetailComposition([fd('fecha', 'text', 'iso_date')]);
      expect(r.slots['subtitle']!.fields, ['fecha']);
    });
  });

  group('buildAutoDetailComposition — stats', () {
    test('un number → stats con 1 field + orientation/separator', () {
      final r = buildAutoDetailComposition([fd('edad', 'number')]);
      expect(r.slots['stats']!.fields, ['edad']);
      expect(r.slots['stats']!.orientation, 'horizontal');
      expect(r.slots['stats']!.separator, ' | ');
    });
    test('5 numbers → primeros 4', () {
      final r = buildAutoDetailComposition([
        fd('a', 'number'), fd('b', 'number'), fd('c', 'number'),
        fd('d', 'number'), fd('e', 'number'),
      ]);
      expect(r.slots['stats']!.fields, ['a', 'b', 'c', 'd']);
    });
    test('cero numbers → sin stats', () {
      final r = buildAutoDetailComposition([fd('nombre', 'text')]);
      expect(r.slots['stats'], isNull);
    });
  });

  group('buildAutoDetailComposition — body / gallery / related', () {
    test('textarea → body', () {
      final r = buildAutoDetailComposition([fd('desc', 'textarea')]);
      expect(r.slots['body']!.fields, ['desc']);
    });
    test('markdown → body', () {
      final r = buildAutoDetailComposition([fd('doc', 'text', 'markdown')]);
      expect(r.slots['body']!.fields, ['doc']);
    });
    test('array<image> → gallery', () {
      final r = buildAutoDetailComposition([fd('fotos', 'array<image>')]);
      expect(r.slots['gallery']!.fields, ['fotos']);
    });
    test('behavior=gallery → gallery', () {
      final r = buildAutoDetailComposition([fd('g', 'array<image>', 'gallery')]);
      expect(r.slots['gallery']!.fields, ['g']);
    });
    test('card_index_list → related', () {
      final r = buildAutoDetailComposition([fd('refs', 'array<number>', 'card_index_list')]);
      expect(r.slots['related']!.fields, ['refs']);
    });
  });

  group('buildAutoDetailComposition — composición completa', () {
    test('sección rica → todos los slots', () {
      final r = buildAutoDetailComposition([
        fd('avatar', 'image', 'avatar'),
        fd('nombre', 'text'),
        fd('year', 'number', 'year'),
        fd('edad', 'number'),
        fd('rating', 'number'),
        fd('desc', 'textarea'),
        fd('fotos', 'array<image>', 'gallery'),
      ]);
      expect(r.recipe, 'hero_protagonico');
      expect(r.slots['avatar']!.fields, ['avatar']);
      expect(r.slots['title']!.fields, ['nombre']);
      expect(r.slots['subtitle']!.fields, ['year']);
      expect(r.slots['stats']!.fields, ['year', 'edad', 'rating']);
      expect(r.slots['body']!.fields, ['desc']);
      expect(r.slots['gallery']!.fields, ['fotos']);
    });
  });

  group('buildAutoListComposition — recipe de LISTA (kind:list, nunca hero)', () {
    test('sección vacía → row_text + slots vacíos + action none', () {
      final r = buildAutoListComposition([]);
      expect(r.recipe, 'row_text');
      expect(r.action, 'none');
      expect(r.slots, isEmpty);
    });

    test('con imagen (avatar) → compact_avatar (NO hero_protagonico)', () {
      final r = buildAutoListComposition([fd('pic', 'image', 'avatar'), fd('nombre', 'text')]);
      expect(r.recipe, 'compact_avatar');
      expect(r.slots['avatar']!.fields, ['pic']);
      expect(r.slots['title']!.fields, ['nombre']);
    });

    test('image plano (sin behavior) también dispara compact_avatar', () {
      final r = buildAutoListComposition([fd('foto', 'image'), fd('nombre', 'text')]);
      expect(r.recipe, 'compact_avatar');
      expect(r.slots['avatar']!.fields, ['foto']);
    });

    test('sin imagen → row_text con title + subtitle', () {
      final r = buildAutoListComposition([fd('nombre', 'text'), fd('anio', 'number', 'year')]);
      expect(r.recipe, 'row_text');
      expect(r.slots['title']!.fields, ['nombre']);
      expect(r.slots['subtitle']!.fields, ['anio']);
    });

    test('NUNCA devuelve una recipe kind:detail', () {
      final r = buildAutoListComposition([
        fd('banner', 'image', 'banner'),
        fd('titulo', 'text'),
        fd('cuerpo', 'textarea'),
      ]);
      expect(['compact_avatar', 'compact_card', 'row_text'], contains(r.recipe));
      expect(r.recipe, isNot('hero_protagonico'));
    });

    test('title y subtitle no colisionan (subtitle != title)', () {
      final r = buildAutoListComposition([fd('nombre', 'text')]);
      expect(r.slots['title']!.fields, ['nombre']);
      expect(r.slots['subtitle'], isNull);
    });
  });
}
